src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,261 @@
+package ui.components.pickers;
+
+import backend.resource.TurboLabel;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used to represent the state of the label picker. In addition to representing a state,
+ * it contains the logic that handles every state transition
+ */
+public class LabelPickerState {
+
+    Set<String> initialLabels;
+    List<String> addedLabels;
+    List<String> removedLabels;
+    List<String> matchedLabels;
+    List<TurboLabel> repoLabels;
+    OptionalInt currentSuggestionIndex;
+
+    @SuppressWarnings("PMD")
+    public LabelPickerState(Set<String> initialLabels, List<TurboLabel> repoLabels, String userInput) {
+        this(initialLabels, new ArrayList<>(), new ArrayList<>(), new ArrayList<>(), repoLabels,
+                OptionalInt.empty());
+        update(userInput);
+    }
+
+    private LabelPickerState(Set<String> initialLabels, List<String> addedLabels, List<String> removedLabels,
+                             List<String> matchedLabels, List<TurboLabel> repoLabels, 
+                             OptionalInt currentSuggestionIndex) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = addedLabels;
+        this.removedLabels = removedLabels;
+        this.matchedLabels = matchedLabels;
+        this.repoLabels = repoLabels;
+        this.currentSuggestionIndex = currentSuggestionIndex;
+    }
+
+    /**
+     * Updates current state based on given user input
+     * @param userInput 
+     */
+    private void update(String userInput) {
+        List<String> confirmedKeywords = getConfirmedKeywords(userInput);
+        for (String confirmedKeyword : confirmedKeywords) {
+            updateIfMatchesOneLabel(confirmedKeyword);
+        }
+
+        Optional<String> keywordInProgess = getKeywordInProgress(userInput);
+        if (keywordInProgess.isPresent()) {
+            updateMatchedLabels(keywordInProgess.get());
+        }
+    }
+
+    /**
+     * Updates current state if there is exactly one matching label based on the 
+     * given keyword
+     * @param keyword
+     */
+    private void updateIfMatchesOneLabel(String keyword) {
+        if (TurboLabel.hasExactlyOneMatchedLabel(repoLabels, keyword)) {
+            updateAssignedLabels(TurboLabel.getFirstMatchingTurboLabel(repoLabels, keyword));
+        }
+    }
```

Comment:
I think we had the confusion because of our previous discussion about `Label` with two words where the user would have to use the mouse to resolve the ambiguity. I guess we can get rid of the `hasExactlyOneMatchedLabel` and `updateIfMatchesOneLabel`.